stock_inventory_non_stocked/models/stock_inventory.py

```diff
@@ -0,0 +1,89 @@
+# Copyright 2024 Ivan Perez <iperez@coninpe.es>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+import ast
+
+from odoo import fields, models
+
+
+class StockInventory(models.Model):
+    _inherit = "stock.inventory"
+
+    create_non_stocked = fields.Boolean(
+        string="Create non stocked products",
+        help="If you select this option, a new quant with zero "
+        "quantity will be created for products without any stock and any quants",
+    )
+
+    def _get_quants(self, locations):
+        res = super(StockInventory, self)._get_quants(locations)
+        if self.create_non_stocked:
+            if self.product_selection == "all":
+                product_ids = self.env["product.product"].search(
+                    [
+                        ("has_quants", "=", False),
+                        ("active", "=", True),
+                        ("type", "=", "product"),
+                    ]
+                )
+                new_quants = self.create_zero_quants(product_ids)
```

Comment:
Maybe you can avoid some code doing one search depending the configuration by building a domain.

Then, with the result, call one 'create_zero_quants()'